test/UnitTests/GitHub.VisualStudio/Services/UsageTrackerTests.cs

```diff
@@ -8,209 +7,346 @@
+            Func<Task> tick = null;
 
-                await target.IncrementCounter(x => x.NumberOfClones);
+            service.WhenForAnyArgs(x => x.StartTimer(null, new TimeSpan(), new TimeSpan()))
+                .Do(x => tick = x.ArgAt<Func<Task>>(0));
+
+            var serviceProvider = CreateServiceProvider();
+            var target = new UsageTracker(serviceProvider, service);
+
+            await tick();
+
+            await service.Received().WriteLocalData(
+                Arg.Is<UsageData>(x =>
+                    x.Model.NumberOfStartups == 0 &&
+                    x.Model.NumberOfStartupsWeek == 0 &&
+                    x.Model.NumberOfStartupsMonth == 0 &&
+                    x.Model.NumberOfClones == 0));
+        }
+
+        [Test]
+        public async Task ShouldIncrementCounter()
+        {
+            var model = new UsageModel { NumberOfClones = 4 };
+            var usageService = CreateUsageService(model);
+            var target = new UsageTracker(
+                CreateServiceProvider(),
+                usageService);
+
+            await target.IncrementCounter(x => x.NumberOfClones);
+            UsageData result = usageService.ReceivedCalls().First(x => x.GetMethodInfo().Name == "WriteLocalData").GetArguments()[0] as UsageData;
+
+            Assert.AreEqual(5, result.Model.NumberOfClones);
+        }
+
+        [Test]
+        public async Task ShouldWriteUpdatedData()
+        {
+            var data = new UsageData { Model = new UsageModel() };
+            var service = CreateUsageService(data);
+            var target = new UsageTracker(
+                CreateServiceProvider(),
+                service);
+
+            await target.IncrementCounter(x => x.NumberOfClones);
+
+            await service.Received(1).WriteLocalData(data);
+        }
+
+        [Test]
+        public async Task UsageServiceWritesAllTheDataCorrectly()
+        {
+            var model = CreateUsageModel();
+            var serviceProvider = CreateServiceProvider();
+            var usageService = CreateUsageService(model, sameDay: true);
+            var targetAndTick = CreateTargetAndGetTick(serviceProvider, usageService);
+            var vsservices = serviceProvider.GetService<IVSServices>();
+            vsservices.VSVersion.Returns(model.VSVersion);
+
+            await targetAndTick.Item2();
+
+            var metricsService = serviceProvider.TryGetService<IMetricsService>();
+
+            var expected = model;
+            expected.NumberOfStartups++;
+            expected.NumberOfStartupsWeek++;
+            expected.NumberOfStartupsMonth++;
+
+            UsageData result = usageService.ReceivedCalls().First(x => x.GetMethodInfo().Name == "WriteLocalData").GetArguments()[0] as UsageData;
+            CollectionAssert.AreEquivalent(
+                ReflectionUtils.GetFields(expected.GetType()).Select(x => new { x.Name, Value = x.GetValue(expected) }),
```

Comment:
Was `SimpleJson.cs` added just for this utility class? Wouldn't `expected.GetType().GetRuntimeFields()` suffice here? I notice you use that method in `CreateUsageModel`.

Reply:
Oh, I must have missed one. Yes, it's for these calls, I don't want to manually type the reflection code, it's really hard to read and easy to miss if you call the wrong one.

Reply:
Sorry I don't understand what you mean by "manually type the reflection code", The two syntaxes are:

`ReflectionUtils.GetFields(expected.GetType())`

or

`expected.GetType().GetRuntimeFields()`

Is the second more "manually typed" than the first? Both are simply calling a method on `expected.GetType()` but the second is less chars as it's an extension method and doesn't require copying `SimpleJson.cs`.

Reply:
I took out all the code that isn't needed in SimpleJson and just left the ReflectionUtils stuff so it doesn't complain about conflicting types.